keps/sig-cli/20190805-kubectl-debug.md

```diff
@@ -0,0 +1,460 @@
+Namespace Sharing] should be enabled for the pod. In cases where process
+namespace sharing isn't enabled for the pod, `kubectl` will set
+`TargetContainer` in the `EphemeralContainer`. This will cause the ephemeral
+container to be created in the namespaces of the target container in runtimes
+that support container namespace targeting.
+
+[Process Namespace Sharing]: https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace
+
+#### Interactive Troubleshooting and Automatic Attaching
+
+Since the primary use case for `kubectl debug` is interactive troubleshooting,
+`kubectl debug` will automatically attach to the console of the newly created
+ephemeral container and will default to creating containers with `Stdin` and
+`TTY` enabled.
+
+These may be disabled via command line flags.
+
+#### Proposed `kubectl debug` arguments
+
+```
+% kubectl help debug
+Execute a container in a pod.
+
+Examples:
+  # Get network address configuration from pod mypod
+  kubectl debug mypod ip addr list
+
+  # Start an interactive debugging session with a debian image
+  kubectl debug mypod --image=debian
+
+  # Run in the same namespace as target container 'myapp'
+  # (Useful for debugging other containers when shareProcessNamespace is false)
+  kubectl debug mypod --target=myapp
+
+Options:
+  -a, --attach=true: Automatically attach to container once created
+  -c, --container='': Container name. If omitted, the first container in the pod will be chosen
+  -i, --stdin=true: Pass stdin to the container
+  --image='': Required. Container image to use for debug container.
+  --target='': Target processes in this container name.
+  -t, --tty=true: Stdin is a TTY
+
+Usage:
+  kubectl debug (POD | TYPE/NAME) [-c CONTAINER] [flags] -- COMMAND [args...] [options]
+
+Use "kubectl options" for a list of global command-line options (applies to all commands).
+```
+
+### Pod Troubleshooting by Copy
+
+Pod troubleshooting via Ephemeral Container relies on an alpha feature which is
+unlikely to be enabled on production clusters. In order to support these use
+cases, and because it would be generally useful, we will support a mode of Pod
+troubleshooting that behaves similar to Pod Troubleshooting with Ephemeral
+Debug Container but operates instead on a copy of the target pod.
+
+The following additional options will cause a copy of the target pod to be
+created:
+
+```
+Options:
+  --copy-to='': Create a copy of the target pod with this pod name.
+  --delete-old: When used with `--copy-to`, delete the original pod.
+  --copy-labels=false: When used with `--copy-to`, specifies whether labels
+                       should also be copied. Note that copying labels may cause
+                       the copy to receive traffic from a service or a replicaset
+                       to kill other pods.
+  --share-processes=true: When used with `--copy-to`, enable process namespace
+                          sharing in the pod copy.
```

Comment:
I'd add `--node-name` to be able to place the pod on a specific node, additionally this will be necessary in the node debug mode.

Reply:
Other options include `--as-user` `--as-root`, which would allow you to run as specific user or as root. `--to-namespace` would be also complementary to `--copy-to`.

Reply:
Generally, we can have several others modifiers for the debug pod, including liveness and readiness probe (you might not care about those), as well as init containers, which might not be needed for debug.

Reply:
Good point about node-name, I'll add that. Or perhaps `--same-node`?

It would be useful to allow arbitrary edits prior to create. What do you think about an `--edit` that opens an editor similar to `kubectl edit`?

Reply:
Re: `--node-name` for node debug mode, I expected the node name to be a positional parameter like `kubectl debug node/$node_name`, but I haven't yet investigated the conventions for specifying resources on the command line.

Reply:
Right, but node-name would be only for node debug, what if you want to land your pod on a specific node for other reasons.